Sources/OktaAnalytics/EventScenario.swift

```diff
@@ -0,0 +1,111 @@
+import Foundation
+import Combine
+
+// Extension to the Date struct, defines the subtraction operator, returning the time interval between two dates
+extension Date {
+    static func - (lhs: Date, rhs: Date) -> TimeInterval {
+        // Returns the time interval between the two dates
+        return lhs.timeIntervalSinceReferenceDate - rhs.timeIntervalSinceReferenceDate
+    }
+}
+
+
+/// This code defines a Swift class `EventScenario` which represents an event that can be tracked with analytics.
+public class EventScenario {
+
+    // eventName is a constant property, initialized with the value passed to the init method
+    let eventName: EventName
+    // properties is a dictionary of strings, initialized as an empty dictionary
+    var properties: Properties = [:]
+    // time is a property of type Date, initialized as the current date
+    var time = Date()
+
+    // eventStream is a PassthroughSubject, which is private set
+    public private(set) var eventStream: PassthroughSubject<Property, Never>?
+
+    // cancellable is an AnyCancellable, which is private
+    private var cancellable: AnyCancellable?
+    // isPaused is a private property which is a boolean, initialized as false
+    private var isPaused = false
+
+    // User defaults object to store properties
+    private let userDefaults: UserDefaults
+
+    // Analytics provider to track events
+    private let analytics: AnalyticsProviderProtocol
+
+    private let eventDateKey: String = "EventDate"
+
+    // Initializer which assigns the passed in name to the eventName property
+    init(_ eventName: EventName, userDefaults: UserDefaults = UserDefaults.standard, analytics: AnalyticsProviderProtocol) {
+        self.eventName = eventName
+        self.userDefaults = userDefaults
+        self.analytics = analytics
+    }
+
+    // Method which starts the event stream, assigns the current time to the time property and sets up a sink to receive values
+    func start() -> PassthroughSubject<Property, Never>? {
+        // check if previous event exists in memory without ended
+        if let eventDate = userDefaults.value(forKey: eventDateKey) as? Date,
+           let properties = userDefaults.value(forKey: eventName) as? Properties,
+           eventDate.distance(to: Date()) > 5.0 /*secs*/ {
+            analytics.trackEvent(eventName, withProperties: properties)
+            userDefaults.removeObject(forKey: eventName)
+        }
+
+        if let eventStream = eventStream {
```

Comment:
If there is one already, if it is 5 minutes and above set scenario_status: Expired and if it is less than 5 minutes, overwritten or interrupted. 
Completed